cooper/formulation/lagrangian_model.py

```diff
@@ -0,0 +1,273 @@
+
+            # We compute (primal) gradients of this object with the sampled
+            # constraints
+            violation = torch.sum(multipliers.detach() * proxy_defect)
+
+            # This is the violation of the "actual/hard" constraint. We use this
+            # to update the multipliers.
+            # The gradients for the dual variables are computed via a backward
+            # on `accumulated_violation_dot_prod`. This enables easy
+            # extensibility to multiplier classes beyond DenseMultiplier.
+
+            # TODO (JGP): Verify that call to backward is general enough for
+            # Lagrange Multiplier models
+            violation_for_update = torch.sum(multipliers * defect.detach())
+            self.update_accumulated_violation(update=violation_for_update)
+
+        return violation
+
+    @no_type_check
+    def backward(
+        self,
+        lagrangian: torch.Tensor,
+        ignore_primal: bool = False,
+        ignore_dual: bool = False,
+    ):
+        """
+        Performs the actual backward computation which populates the gradients
+        for the primal and dual variables.
+
+        Args:
+            lagrangian: Value of the computed Lagrangian based on which the
+                gradients for the primal and dual variables are populated.
+            ignore_primal: If ``True``, only the gradients with respect to the
+                dual variables are populated (these correspond to the constraint
+                violations). This feature is mainly used in conjunction with
+                ``alternating`` updates, which require updating the multipliers
+                based on the constraints violation *after* having updated the
+                primal parameters. Defaults to False.
+            ignore_dual: If ``True``, the gradients with respect to the dual
+                variables are not populated.
+        """
+
+        if ignore_primal:
+            # Only compute gradients wrt Lagrange multipliers
+            # No need to call backward on Lagrangian as the dual variables have
+            # been detached when computing the `weighted_violation`s
+            pass
+        else:
+            # Compute gradients wrt _primal_ parameters only.
+            # The gradient for the dual variables is computed based on the
+            # non-proxy violations below.
+            lagrangian.backward()
+
+        # Fill in the gradients for the dual variables based on the violation of
+        # the non-proxy constraints
+        if not ignore_dual:
+            dual_vars = self.dual_parameters
+            self.accumulated_violation_dot_prod.backward(inputs=dual_vars)
+            # FIXME(IsitaRex): Alternative implementation had the code below
+            # It's unclear why this difference exists. Must be checked before
+            # merging into main
+            # new_loss = lagrangian - self.accumulated_violation_dot_prod
+            # new_loss.backward(inputs=dual_vars)
```

Comment:
This alternative implementation should be disregarded, that implementation corresponds to confusion made while looking at the authors' code, it was a test. The current implementation is the one that works and should be kept.